src/shared/passwdqc-dlopen.c

```diff
@@ -0,0 +1,137 @@
+char *(*sym_passwdqc_random)(const passwdqc_params_t *params);
+
+int dlopen_passwdqc(void) {
+        return dlopen_many_sym_or_warn(
+                        &passwdqc_dl, "libpasswdqc.so.1", LOG_DEBUG,
+                        DLSYM_ARG(passwdqc_params_reset),
+                        DLSYM_ARG(passwdqc_params_load),
+                        DLSYM_ARG(passwdqc_params_parse),
+                        DLSYM_ARG(passwdqc_params_free),
+                        DLSYM_ARG(passwdqc_check),
+                        DLSYM_ARG(passwdqc_random));
+}
+
+#define PASSWDQC_CONFIG "/etc/passwdqc.conf"
+
+static int pwqc_allocate_context(passwdqc_params_t **ret) {
+        _cleanup_(sym_passwdqc_params_freep) passwdqc_params_t *params = NULL;
+        char *load_reason;
+        int r;
+
+        assert(ret);
+
+        r = dlopen_passwdqc();
+        if (r < 0)
+                return r;
+
+        params = malloc(sizeof(passwdqc_params_t));
+        if (params == NULL)
+                return -errno;
+
+        sym_passwdqc_params_reset(params);
+
+        r = sym_passwdqc_params_load(params, &load_reason, PASSWDQC_CONFIG);
+        if (r < 0) {
+                log_debug("Failed to load passwdqc configuration '%s', ignoring: %s",
+                          PASSWDQC_CONFIG, load_reason);
+                free(load_reason);
+        }
+
+        *ret = TAKE_PTR(params);
+        return 0;
+}
+
+#define N_SUGGESTIONS 6
+
+int suggest_passwords(void) {
+        _cleanup_(sym_passwdqc_params_freep) passwdqc_params_t *params = NULL;
+        _cleanup_strv_free_erase_ char **suggestions = NULL;
+        _cleanup_(erase_and_freep) char *joined = NULL;
+        size_t i;
+        int r;
+
+        r = pwqc_allocate_context(&params);
+        if (ERRNO_IS_NOT_SUPPORTED(r))
+                return 0;
+        if (r < 0)
+                return log_error_errno(r, "Failed to allocate libpasswdqc context: %m");
+
+        suggestions = new0(char*, N_SUGGESTIONS+1);
+        if (!suggestions)
+                return log_oom();
+
+        for (i = 0; i < N_SUGGESTIONS; i++) {
+                suggestions[i] = sym_passwdqc_random(params);
+                if (suggestions[i] == NULL)
+                        return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to generate password, ignoring");
+        }
+
+        joined = strv_join(suggestions, " ");
+        if (!joined)
+                return log_oom();
+
+        log_info("Password suggestions: %s", joined);
+        return 1;
+}
+
+int check_password_quality(
+                const char *password,
+                const char *old,
+                const char *username,
+                char **ret_error) {
+        _cleanup_(sym_passwdqc_params_freep) passwdqc_params_t *params = NULL;
+        const char *check_reason;
+        const struct passwd pw = {
+                .pw_name = username,
+                .pw_gecos = "",
+                .pw_dir = ""};
```

Comment:
indentation off.

Why is the rest of the fields not initialized?

How does this even compile? the fields are not const but the strings are

Reply:
> Why is the rest of the fields not initialized?

passwdqc requires this 3 field to be not NULL, but we use only username. 

> How does this even compile? the fields are not const but the strings are

Made strings not const either.